crypto/ts/ts_rsp_verify.c

```diff
@@ -198,52 +198,58 @@ static int ts_check_signing_certs(PKCS7_SIGNER_INFO *si,
                                   STACK_OF(X509) *chain)
 {
     ESS_SIGNING_CERT *ss = ossl_ess_signing_cert_get(si);
-    STACK_OF(ESS_CERT_ID) *cert_ids = NULL;
     ESS_SIGNING_CERT_V2 *ssv2 = ossl_ess_signing_cert_v2_get(si);
-    STACK_OF(ESS_CERT_ID_V2) *cert_ids_v2 = NULL;
     X509 *cert;
     int i = 0;
     int ret = 0;
+    int id_index;
+    int chain_length;
+    int verification_length;
+    void *cert_ids_generic;
+    int (*find_cert_generic)(const void *, const X509 *);
 
     if (ss != NULL) {
-        cert_ids = ss->cert_ids;
-        cert = sk_X509_value(chain, 0);
-        if (ossl_ess_find_cert(cert_ids, cert) != 0)
-            goto err;
-
-        /*
-         * Check the other certificates of the chain if there are more than one
-         * certificate ids in cert_ids.
-         */
-        if (sk_ESS_CERT_ID_num(cert_ids) > 1) {
-            for (i = 1; i < sk_X509_num(chain); ++i) {
-                cert = sk_X509_value(chain, i);
-                if (ossl_ess_find_cert(cert_ids, cert) < 0)
-                    goto err;
-            }
-        }
+        cert_ids_generic = ss->cert_ids;
+        find_cert_generic = (int (*)(const void *, const X509 *))&ossl_ess_find_cert;
+        verification_length = sk_ESS_CERT_ID_num(cert_ids_generic);
     } else if (ssv2 != NULL) {
-        cert_ids_v2 = ssv2->cert_ids;
-        cert = sk_X509_value(chain, 0);
-        if (ossl_ess_find_cert_v2(cert_ids_v2, cert) != 0)
-            goto err;
-
-        /*
-         * Check the other certificates of the chain if there are more than one
-         * certificate ids in cert_ids.
-         */
-        if (sk_ESS_CERT_ID_V2_num(cert_ids_v2) > 1) {
-            for (i = 1; i < sk_X509_num(chain); ++i) {
-                cert = sk_X509_value(chain, i);
-                if (ossl_ess_find_cert_v2(cert_ids_v2, cert) < 0)
-                    goto err;
-            }
-        }
+        cert_ids_generic = ssv2->cert_ids;
+        find_cert_generic = (int (*)(const void *, const X509 *))&ossl_ess_find_cert_v2;
+        verification_length = sk_ESS_CERT_ID_V2_num(cert_ids_generic);
     } else {
         goto err;
     }
 
+    if (verification_length < 1) /* no signer's ESSCertID */
+        goto err;
+
+    chain_length = sk_X509_num(chain);
+    if (chain_length < 1) /* empty chain */
+        goto err;
+
+    /* RFC 2634, section 5.4 */
+    if (1 < verification_length) {
+        if (1 < chain_length)
```

Comment:
We usually put constant on the right hand side of the condition.

Reply:
I always use `<` where possible; the objective is to represent the number line.  
I found it less prone to error this way because people can imagine the number line more easily. And having the mental visual representation helps a lot.

I found a similar order in the same file, so I guessed someone likes it too.
https://github.com/openssl/openssl/blob/a368e4/crypto/ts/ts_rsp_verify.c#L381
Anyway, let's do it as suggested, I only want to explain the rationale behind it.